src/serde_xml_plist.rs

```diff
@@ -1,467 +1,657 @@
-//! Helpers for deserializing plist values from untyped XML
+//! Helpers for (de)serializing plist values from untyped XML
 //!
 //! This is essentially glue between the `plist` crate and the `quick_xml` crate.
 //! It allows plist values, dictionaries and arrays to be used inside types that
 //! derive Deserialize/Serialize.
 
-use std::{fmt::Display, marker::PhantomData, str::FromStr};
-
 use base64::{engine::general_purpose::STANDARD as base64_standard, Engine};
 use plist::{Dictionary, Value};
-use serde::{
-    de::{Error as DeError, Visitor},
-    Deserialize, Deserializer,
-};
+use serde::ser::SerializeStruct;
+use serde::{Deserializer, Serializer};
 
 /// Deserialize a plist Dictionary
 ///
 /// This relies on the specific structure presented by the quick_xml crate and
 /// is likely not suited to other formats.
-pub fn deserialize_dict<'de, D>(deserializer: D) -> Result<Dictionary, D::Error>
+pub(crate) fn deserialize<'de, D>(deserializer: D) -> Result<Dictionary, D::Error>
 where
     D: Deserializer<'de>,
 {
-    deserializer.deserialize_map(ValueVisitor::DictOnly).map(|x| x.into_dictionary().unwrap())
+    deserializer.deserialize_map(de::ValueVisitor::DictOnly).map(|x| x.into_dictionary().unwrap())
 }
 
-struct DictWrapper(Dictionary);
-struct ValueWrapper(Value);
-struct ArrayWrapper(Vec<Value>);
-struct IntWrapper(plist::Integer);
-
-/// The literal keyword 'key'.
-struct KeyKeywordLiteral;
+mod de {
+    use super::*;
+    use serde::de::{Error as DeError, Visitor};
+    use serde::Deserialize;
+    use std::fmt::Display;
+    use std::marker::PhantomData;
+    use std::str::FromStr;
+
+    struct DictWrapper(Dictionary);
+
+    struct ValueWrapper(Value);
+
+    struct ArrayWrapper(Vec<Value>);
+
+    struct IntWrapper(plist::Integer);
+
+    /// The literal keyword 'key'.
+    struct KeyKeywordLiteral;
+
+    /// PLIST value keywords
+    ///
+    /// We use types for keywords, with custom deserialize impls, to avoid needing
+    /// to transiently allocate strings each time we encounter them. :shrug:
+    enum ValueKeyword {
+        Dict,
+        Array,
+        Integer,
+        Real,
+        String,
+        Data,
+        Date,
+        True,
+        False,
+    }
 
-/// PLIST value keywords
-///
-/// We use types for keywords, with custom deserialize impls, to avoid needing
-/// to transiently allocate strings each time we encounter them. :shrug:
-enum ValueKeyword {
-    Dict,
-    Array,
-    Integer,
-    Real,
-    String,
-    Data,
-    Date,
-    True,
-    False,
-}
+    // the logic for deserializing a dict is a subset of the general deser logic,
+    // so we reuse this type for both cases.
+    pub(super) enum ValueVisitor {
```

Comment:
if this is shared between ser&de, I would just move it into the root module?

Reply:
Ah I assumed that the `pub(super)` was there so that this could be shared, but I think I just misunderstood, no worries. I'm not sure unifying is possible, and I'm not too worried about it as long as we have something that works.